### ISSUE TITLE
Delayed start possible? NTP: Time set after startup

### ISSUE THREAD

**Issue by DonHansDampf** [quantum-elixir/quantum-core#395]
Greetings!

I would like to use your module on some elixir nerves projects of mine. Sadly it seems like there are some problems in embedded cases like this: Another elixir module (`nerves-time`) updates the system time using ntp. There was already another ticket open with this issue: https://github.com/quantum-elixir/quantum-core/issues/342

In my case the startup order was a bit different. It seemed like Quantum already used the new time since the next Execution was scheduled for the correct time (not in the year 1970). But sill the job was never executed. My guess is, that quantum scheduled the task for example 08:30:01 at the current time 08:29:57. Right after that NTP corrects the a bit and Quantum never reaches the point 08:30:01 again since it's already 08:30:12. 

I think i am not the only one who wants to use your awesome module in a embedded project. So, is there some way to delay the start of quantum?

Another possibility could be delay Quantum till a function returns true. In nerves case `Nerves.Time.synchronized?/0` could be used to check the current "state" of time-sync. 

I think if you don't want to make any changes to your codebase, another more brutal solution would be, to create another GenServer, which checks the state of ntp with the given function. After the function returns true, use `Supervisor.restart_child`. But it would be cleaner without a workaround like that.

Hope you can help me out, thanks!

**Comment by maennchen** [comment]
Quantum currently assumes time that moves forward. It should however not skip entries.

Currently, it works like this: The next execution date is determined for every job. The jobs are then grouped and ordered by that date. A timer is started for the time of the first job (the next one to execute). When this timer fires, the job is started, the time of the Scheduler advanced to the execution time + 1s. The process is then started over.

Given that functionality, jobs should be executed also when the time has already passed. It will however not re-calculate the timer when the time changes.

Would it be a possibility to have a function to call when the time is changed to recalculate all the timers?

**Comment by DonHansDampf** [comment]
Thank you for the explanation. Helped me to understand some underlying groundwork which gets executed! :)

> Would it be a possibility to have a function to call when the time is changed to recalculate all the timers?

Sounds like a good idea. Some kind of reload function which can be used. I think in most cases the developer himself knows when time jumps occur (for example in embedded projects like mine) and could create his own hook to execute the reload. Or use the function in some module's init function till the time is corrected (`Nerves.Time.synchronized?/0`). Would be more graceful as killing the Scheduler module and letting the supervisor restart its child.

**Comment by AdrianRibao** [comment]
I've replaced `nerves_ntp` by `nerves_time` and it's better now, tasks won't run thousands of time.

Now they just won't run, but if I stop the scheduler manually, then it works as expected.

I could create a genserver that stops the scheduller after X minutes since initialization, but that is a very ugly hack. Do you recommend a better way to make it work?

This is related to issue #395. Any progress on that?

**Comment by cblavier** [comment]
I have the same issue (working on a nerves project that needs NTP to be in sync before working). I wrote a custom supervisor that don't run anything until NTP is in sync, but when my supervisor finally start my Scheduler (a few seconds after NTP is in sync), my jobs never get triggered.

If I manually terminate and restart my Scheduler process in iex, it will eventually work.
Any idea of what I could have missed?

**Comment by maennchen** [comment]
@cblavier Not really. Is there any way how I could debug this myself?

Can you post the debug logs?

**Comment by cblavier** [comment]
Sure, here are my debug logs at the moment I start the scheduler:

```
10:52:30.783 [info]  ntpd synchronization changed (now true): %{freq_drift_ppm: 0, offset: -0.001935, poll_interval: 32, stratum: 3}
10:52:31.569 [info]  NTP time is now synchronized
10:52:36.569 [info]  NTPSupervisor restarting children
10:52:36.570 [info]  Starting {PhenixPlayer.DBus.DBusServer, [implementation: PhenixPlayer.DBus.DBusRpi3Impl]}
10:52:36.663 [info]  Starting {PhenixPlayer.MediaDownloader, []}
10:52:36.697 [info]  Starting {PhenixPlayer.Scheduler, []}
10:52:36.700 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.JobBroadcaster] Loading Initial Jobs from Config
10:52:36.701 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.ExecutionBroadcaster] Unknown last execution time, using now
10:52:36.703 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.ExecutionBroadcaster] Adding job #Reference<0.1356365699.536870915.12960>
10:52:36.703 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.ExecutionBroadcaster] Adding job #Reference<0.1356365699.536870915.12970>
10:52:36.704 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.ExecutionBroadcaster] Continuing Execution Broadcasting at 1561459959995 (2019-06-25T10:52:37)
10:52:36.705 [info]  Terminating PhenixPlayer.NTP.NTPMonitor
```

Then none of my 2 scheduled jobs (one every second, the other every 10 second is executed)
As soon as I run:

```
[{job1, _}, {job2, _}] = PhenixPlayer.Scheduler.jobs
PhenixPlayer.Scheduler.deactivate_job(job1)
```
It unblocks the scheduler and my second job runs again (and I activate job1 again it is scheduled flawlessly)

```
10:53:09.900 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.JobBroadcaster] Change job state #Reference<0.1356365699.536870915.12960>
10:53:09.901 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.ExecutionBroadcaster] Removing job #Reference<0.1356365699.536870915.12960>
10:53:09.901 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.ExecutionBroadcaster] Continuing Execution Broadcasting ASAP (2019-06-25T10:52:40)
10:53:09.902 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.ExecutionBroadcaster] Scheduling job for execution #Reference<0.1356365699.536870915.12970>
10:53:09.903 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.ExecutionBroadcaster] Continuing Execution Broadcasting ASAP (2019-06-25T10:52:50)
10:53:09.903 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.ExecutionBroadcaster] Scheduling job for execution #Reference<0.1356365699.536870915.12970>
10:53:09.904 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.ExecutionBroadcaster] Continuing Execution Broadcasting ASAP (2019-06-25T10:53:00)
10:53:09.905 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.ExecutionBroadcaster] Scheduling job for execution #Reference<0.1356365699.536870915.12970>
10:53:09.906 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.Executor] Task for job #Reference<0.1356365699.536870915.12970> started on node :"phenix_player@nerves.local"
10:53:09.906 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.ExecutionBroadcaster] Continuing Execution Broadcasting at 96294 (2019-06-25T10:53:10)
10:53:09.907 [debug] [:"phenix_player@nerves.local"][Elixir.Quantum.Executor] Execute started for job #Reference<0.1356365699.536870915.12970>
```

**Comment by maennchen** [comment]
Nothing comes to mind with those logs. I'd actually need to `IO.inspect` debug this one I guess...

**Comment by cblavier** [comment]
It's not easy for me let you run the code as you'd need to run it on a raspberry.
Maybe you could pinpoint me in your codebase where I should `IO.inspect` things?

**Comment by maennchen** [comment]
I'd mostly check if the `ExecutionBroadcaster` got the events and its internal state when it is changed.

The most important part is that the `ExecutionBroadcaster` sets up some timers to actually trigger execution.

**Comment by maennchen** [comment]
@cblavier Do you have any more information? Otherwise I'm closing the issue.

**Comment by cblavier** [comment]
Please close! Thank you
